boards: renesas: Add support for Renesas EK-RA4M1

quytranpzz opened the pull request:
- Add support for EK-RA4M1 board 
- Add device tree support for EK-RA4M1 with basic gpio, pinctrl, uart, clock control, counter, spi and adc drivers.
- Add SoC support for `r7fa4m1ab3cfp`

quytranpzz committed:
boards: renesas: Add support for Renesas EK-RA4M1 board

Add minimal support for Renesas EK-RA4M1 board with
basic gpio, pinctrl, uart, clock control and counter
drivers"

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

quytranpzz committed:
dts: renesas: Add device tree support for Renesas RA4M1 SoC

- Add device tree support for Renesas r7fa4m1ab3cfp with basic
gpio, clock control, sci, counter drivers, adc
- Add clock-frequency in renesas,ra-cgc-pclk binding to use the
clock frequency for iclk

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

quytranpzz committed:
soc: renesas: ra: Add SoC support for Renesas RA4M1

Add support for Renesas RA4M1 SoC series with r7fa4m1ab3cfp

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

quytranpzz committed:
boards: renesas: Add board device tree support for SPI on EK-RA4M1

- Add board devicetree support for SPI driver on Renesas EK-RA4M1

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

nordicjm commented on boards/renesas/ek_ra4m1/ek_ra4m1_defconfig:
```diff
@@ -0,0 +1,17 @@
+# Copyright (c) 2024 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=48000000
```
shouldn't this be set by SoC?

quytranpzz replied:
I believe it is more reasonable to set this value on board, as it will depend on the actual clock supply available on the boards.

nordicjm replied:
@gmarull

gmarull replied:
this likely means you need to take this from DT?

quytranpzz replied:
sorry for late response,
I moved it into SoC and took it from DT, thank you

nordicjm reviewed (commented):
Board files OK, one comment about a Kconfig

gmarull commented on boards/renesas/ek_ra4m1/ek_ra4m1_defconfig:
```diff
@@ -0,0 +1,17 @@
+CONFIG_UART_CONSOLE=y
+CONFIG_CONSOLE=y
+
+CONFIG_BUILD_OUTPUT_HEX=y
+CONFIG_BUILD_NO_GAP_FILL=y
+CONFIG_CLOCK_CONTROL=y
```
these are not board level configs

thaoluonguw replied:
@gmarull : Thank you for your feedback. May I confirm that which level is suitable for these configs? (ex. CONFIG_CLOCK_CONTROL should be in SoC level, CONFIG_BUILD_OUTPUT_HEX ?).

quytranpzz replied:
I have moved it into SoC configs

thaoluonguw commented on dts/arm/renesas/ra/ra4/r7fa4m1ab3cfp.dtsi:
```diff
@@ -0,0 +1,190 @@
+			reg = <0x20000000 DT_SIZE_K(32)>;
+		};
+
+		flash-controller@407e0000 {
+			reg = <0x407e0000 0x1000>;
```
It should be "0x407E**C**000"
![image](https://github.com/user-attachments/assets/6ef9b74b-7af6-46cb-8445-1e0595da08b8)

quytranpzz replied:
In the current flash driver development, I use the full Flash I/O register as mentioned in the Address space ![image](https://github.com/user-attachments/assets/64893ef0-4d26-40d1-9053-b54be1fd467f)

quytranpzz commented on boards/renesas/ek_ra4m1/doc/index.rst:
```diff
@@ -0,0 +1,167 @@
+- MCU reset push-button switch
+- MCU boot configuration jumper
+
+**Special Feature Access**
```
I removed it, thank you

quytranpzz commented:
Hi @KhiemNguyenT , @soburi , @thaoluonguw , @duynguyenxa : please take a look

quytranpzz committed:
boards: renesas: Add board device tree support for ADC on EK-RA4M1

Add dts property support for adc on ek_ra4m1

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

quytranpzz committed:
dts: renesas: ra: Add support for gpio interrupt

Add support for gpio interrupt pins for r7fa4m1ab3cfp

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

quytranpzz committed:
board: renesas: Add support for on-board button on EK-RA4M1

Add gpio on-board key support for ek_ra4m1

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

quytranpzz committed:
dts: renesas: Update dts property for ra4-cm4-common

- Remove redundant node adc1 on ra4-cm4-common due to unsupported
- Update interrupt number for spi1

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

nordicjm commented on drivers/spi/Kconfig.renesas_ra:
```diff
@@ -28,8 +28,8 @@ config SPI_RA_DTC
 config SPI_USE_HW_SS
 	bool "RA MCU SPI Hardware Slave Select support"
 	default y
-	depends on !SOC_SERIES_RA2A1
+	depends on !(SOC_SERIES_RA2A1 || SOC_SERIES_RA4M1)
 	help
-	  Use Slave Select pin instead of software Slave Select.
+	  Use Hardware Slave Select (SSL) instead of Software Slave Select.
```
update as per https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-2-inclusive-language

quytranpzz replied:
updated, thank you

quytranpzz commented:
Add ADC and GPIO interrupt for adapting the merged PRs #80719 and #79849

quytranpzz commented:
Rebases to solve conflicts

KhiemNguyenT reviewed (dismissed):
LGTM

kartben commented on drivers/spi/Kconfig.renesas_ra:
```diff
@@ -26,10 +26,10 @@ config SPI_RA_DTC
+	depends on !(SOC_SERIES_RA2A1 || SOC_SERIES_RA4M1)
 	help
-	  Use Slave Select pin instead of software Slave Select.
+	  Use Hardware Peripheral Select (SSL) instead of Software Peripheral Select.
```
SSL acronym doesn't seem relevant/accurate anymore

quytranpzz committed:
drivers: clock_control: Add condition to verify the CPU clock config

- Add a condition to check the clock supplying the CPU to match with
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
- Correct CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC for EK-RA4W1

Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

quytranpzz commented:
Adds commit to verify the clock configuration

KhiemNguyenT reviewed (dismissed):
Thank you. LGTM.

quytranpzz commented:
Hi @gmarull , @kartben : please take a look. Thank you.

kartben commented on boards/renesas/ek_ra4m1/doc/index.rst:
```diff
@@ -0,0 +1,159 @@
+
+Hardware
+********
+Detailed hardware feature for the RA4M1 MCU group can be found at `RA4M1 Group User's Manual Hardware`_
```
```suggestion
Detailed hardware features for the RA4M1 MCU group can be found at `RA4M1 Group User's Manual Hardware`_
```

quytranpzz replied:
updated, thank you

kartben commented on boards/renesas/ek_ra4m1/doc/index.rst:
```diff
@@ -0,0 +1,159 @@
+********
+Detailed hardware feature for the RA4M1 MCU group can be found at `RA4M1 Group User's Manual Hardware`_
+
+.. figure:: ra4m1_block_diagram.webp
+	:width: 442px
+	:align: center
+	:alt: RA4M1 MCU group feature
+
+	RA4M1 Block diagram (Credit: Renesas Electronics Corporation)
+
```
please drop this and the associated image file ; you're already cross-referencing the MCU and board's manual so there's no need for a new 50K image resource in the git repo for something that people really interested can learn more about in the linked documents.

quytranpzz replied:
I dropped it, and the associated image file as well. thank you.

kartben commented on boards/renesas/ek_ra4m1/doc/index.rst:
```diff
@@ -0,0 +1,159 @@
+
+	RA4M1 Block diagram (Credit: Renesas Electronics Corporation)
+
+Detailed hardware feature for the EK-RA4M1 MCU can be found at `EK-RA4M1 - User's Manual`_
```
```suggestion
Detailed hardware features for the EK-RA4M1 MCU can be found at `EK-RA4M1 - User's Manual`_
```

quytranpzz replied:
fixed, thank you

kartben commented on boards/renesas/ek_ra4m1/doc/index.rst:
```diff
@@ -0,0 +1,159 @@
+
+To flash the program to board
+
+  1. Connect to J-Link OB via USB port to host PC
+
+  2. Make sure J-Link OB jumper is in default configuration as describe in `EK-RA4M1 - User's Manual`_
+
+  3. Execute west command
+
+	.. code-block:: console
+
+		west flash -r jlink
```
```suggestion
1. Connect to J-Link OB via USB port to host PC

2. Make sure J-Link OB jumper is in default configuration as describe in `EK-RA4M1 - User's Manual`_

3. Execute west command

   .. code-block:: console

      west flash -r jlink
```

kartben commented on boards/renesas/ek_ra4m1/doc/index.rst:
```diff
@@ -0,0 +1,154 @@
+
+2. Make sure J-Link OB jumper is in default configuration as describe in `EK-RA4M1 - User's Manual`_
+
+3. Execute west command
+
+.. code-block:: console
+
+	west flash -r jlink
```
doesn't look like you applied my suggested change correctly?


```suggestion
3. Execute west command

   .. code-block:: console

      west flash -r jlink
```

quytranpzz replied:
sorry for the mistake. I corrected it.

nordicjm commented on boards/renesas/ek_ra4m1/doc/index.rst:
```diff
@@ -0,0 +1,154 @@
+Supported Features
+==================
+
+The below features are currently supported on Zephyr OS for EK-RA4M1 board:
```
```suggestion
The below features are currently supported on Zephyr for the ``ek_ra4m1`` board:
```

quytranpzz replied:
updated, thanks.

nordicjm commented on dts/arm/renesas/ra/ra4/r7fa4m1ab3cfp.dtsi:
```diff
@@ -0,0 +1,190 @@
+			reg = <0x40047000 4>,
+				  <0x40047004 4>,
+				  <0x40047008 4>;
+			reg-names = "MSTPB","MSTPC",
+						"MSTPD";
```
one line?

quytranpzz replied:
updated to one line

nordicjm commented on dts/arm/renesas/ra/ra4/r7fa4m1ab3cfp.dtsi:
```diff
@@ -0,0 +1,190 @@
+		pll: pll {
+			compatible = "renesas,ra-cgc-pll";
+			#clock-cells = <0>;
+
+			/* PLL */
```
comment seems unnecessary
```suggestion
```

quytranpzz replied:
Removed comment

nordicjm commented on dts/arm/renesas/ra/ra4/r7fa4m1ab3cfp.dtsi:
```diff
@@ -0,0 +1,190 @@
+			reg = <0x40070040 0x20>;
+			clocks = <&pclka MSTPB 29>;
+			status = "disabled";
+			uart {
```
newline

quytranpzz replied:
Added new line, thanks

nordicjm commented on boards/renesas/ek_ra4m1/doc/index.rst:
```diff
@@ -0,0 +1,154 @@
+Programming and Debugging
+*************************
+
+Applications for the ``ek_ra4m1`` board target configuration can be
```
```suggestion
Applications for the ``ek_ra4m1`` board can be
```

quytranpzz replied:
updated. thanks

nordicjm commented on boards/renesas/ek_ra4m1/ek_ra4m1.dts:
```diff
@@ -0,0 +1,66 @@
+
+	leds {
+		compatible = "gpio-leds";
+		led1: led1 {
```
inconsistent spacing

quytranpzz replied:
I updated, thanks

nordicjm commented on boards/renesas/ek_ra4m1/ek_ra4m1.dts:
```diff
@@ -0,0 +1,66 @@
+	pinctrl-0 = <&sci1_default>;
+	pinctrl-names = "default";
+	status = "okay";
+	uart1: uart {
```
and here

quytranpzz replied:
I updated as well, thanks

nordicjm commented on drivers/clock_control/clock_control_renesas_ra_cgc.c:
```diff
@@ -23,6 +23,20 @@ static volatile uint32_t *mstp_regs[] = {
 static volatile uint32_t *mstp_regs[] = {};
 #endif
 
+#if !CONFIG_PM
```
```suggestion
#if !defined(CONFIG_PM)
```

quytranpzz replied:
I updated, thanks

quytranpzz commented:
@kartben , @nordicjm : Thanks very much for your review. I have updated the mistakes. Please take a look.

nordicjm commented on dts/arm/renesas/ra/ra4/r7fa4m1ab3cfp.dtsi:
```diff
@@ -0,0 +1,189 @@
+			reg = <0x407e0000 0x10000>;
+			#address-cells = <1>;
+			#size-cells = <1>;
+			flash0: flash@0 {
```
newline for consistency

quytranpzz replied:
updated, thank you

nordicjm commented on boards/renesas/ek_ra4m1/ek_ra4m1.dts:
```diff
@@ -30,8 +31,18 @@
 
+	buttons {
+		compatible = "gpio-keys";
+		button0: s1 {
```
newline for consistency

quytranpzz replied:
updated, thank you

nordicjm reviewed (commented):
other files look OK
